### PR TITLE
Refactor HostIP predicate algorithm

### PR DESCRIPTION
**Pull request by anfernee**
- Remove string decode logic. It's not really helping to find the
  conflict ports, and it's expensive to do encoding/decoding
- Not to parse the container ports information in predicate meta, use
  straight []*v1.ContainerPort
- Use better data structure to search port conflict based on ip
  addresses
- Collect scattered source code into common place
- related to #52421

**Special notes for your reviewer**: @k82cn 

**Release note**:
```release-note
None
```

### CODE REVIEW

File: plugin/pkg/scheduler/util/utils.go [hunk 1]
```diff
@@ -28,33 +27,115 @@ import (
 
 const DefaultBindAllHostIP = "0.0.0.0"
 
-// GetUsedPorts returns the used host ports of Pods: if 'port' was used, a 'port:true' pair
+type ProtocolPort struct {
```
Comment (ravisantoshgudimetla) [inline plugin/pkg/scheduler/util/utils.go]: Can you add comments here for both functions and types. The changes look good. Wondering why golint isn't complaining.

Comment (resouer) [reply plugin/pkg/scheduler/util/utils.go]: Is it true that golint is enforced in CI?

**Review by bsalamat — COMMENTED** [review]
Thanks, @anfernee and sorry for the delay in reviewing this PR. I have a few minor comments.

File: plugin/pkg/scheduler/util/utils.go (hunk 1 above)
Comment (anfernee) [reply plugin/pkg/scheduler/util/utils.go]: right, I don't think CI checks it.. will add them.

File: pkg/scheduler/util/utils_test.go [hunk 2]
```diff
@@ -93,3 +93,209 @@ func TestSortableList(t *testing.T) {
...
+			desc: "empty ip and protocol add should work",
+			added: []hostPortInfoParam{
+				{"", "127.0.0.1", 79},
+				{"UDP", "127.0.0.1", 80},
+				{"", "127.0.0.1", 81},
+				{"", "127.0.0.1", 82},
+				{"", "", 79},
+				{"UDP", "", 80},
+				{"", "", 81},
+				{"", "", 82},
```
Comment (bsalamat) [inline pkg/scheduler/util/utils_test.go]: Just out of curiosity, what will happen if someone provides a negative port number?

Comment (bsalamat) [reply pkg/scheduler/util/utils_test.go]: My question is not necessarily related to this line of code. It is a general question.

Comment (anfernee) [reply pkg/scheduler/util/utils_test.go]: scheduler will reserve the negative port as well. however, the api ensures the ports are always positive, so it should be good. but you are right, scheduler can be extra cautious about this kind of input. 

```golang
	// Number of port to expose on the host.
	// If specified, this must be a valid port number, 0 < x < 65536.
	// If HostNetwork is specified, this must match ContainerPort.
	// Most containers do not need this.
	// +optional
	HostPort int32 `json:"hostPort,omitempty" protobuf:"varint,2,opt,name=hostPort"`
	// Number of port to expose on the pod's IP address.
	// This must be a valid port number, 0 < x < 65536.
	ContainerPort int32 `json:"containerPort" protobuf:"varint,3,opt,name=containerPort"`
```
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L1742-L1750

**Comment by bsalamat** [comment]
I just have a minor question about validating port. Otherwise, it LGTM.

File: pkg/scheduler/util/utils_test.go (hunk 2 above)
Comment (bsalamat) [reply pkg/scheduler/util/utils_test.go]: You are right. K8s API ensure that the port number is valid:
https://github.com/bsalamat/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go#L201

So, I think we don't need to repeat the validation in scheduler code.

**Comment by bsalamat** [comment]
Thanks, @anfernee!

/lgtm

**Pull request merged**